src/Sound/Tidal/Stream.hs

```diff
@@ -426,62 +494,65 @@ streamFirst stream pat = do now <- O.time
-         -- If a 'fake' tick, it'll be aligned with cycle zero
-         pat | fake = withResultTime (+ cycleNow) patstack
-             | otherwise = patstack
-         frameEnd = snd $ T.nowTimespan st
-         -- add cps to state
-         sMap' = Map.insert "_cps" (VF (T.cps tempo)) sMap
-         --filterOns = filter eventHasOnset
-         extraLatency | fake = 0
-                      | otherwise = cFrameTimespan config + T.nudged tempo
-         -- First the state is used to query the pattern
-         es = sortOn (start . part) $ query pat (State {arc = T.nowArc st,
+    return sMap) (do
+      pMap <- readMVar (sPMapMV stream)
+      busses <- readMVar (sBusses stream)
+      sGlobalF <- readMVar (sGlobalFMV stream)
+      bpm <- (T.getTempo ops)
+      cycleNow <- (T.timeToCycles ops) $ T.start st
+      let
+        config = sConfig stream
+        cxs = sCxs stream
+        patstack = sGlobalF $ playStack pMap
+        -- If a 'fake' tick, it'll be aligned with cycle zero
+        pat | fake = withResultTime (+ cycleNow) patstack
+            | otherwise = patstack
+        frameEnd :: Link.Micros
+        frameEnd = T.nowEnd st
+        -- add cps to 
+        cps = (T.beatToCycles ops) bpm
+        sMap' = Map.insert "_cps" (VF $ coerce cps) sMap
+        --filterOns = filter eventHasOnset
+        extraLatency | fake = 0
+                     | otherwise = T.nudged st
+        -- First the state is used to query the pattern
+        es = sortOn (start . part) $ query pat (State {arc = T.nowArc st,
                                                         controls = sMap'
-                                                       }
+                                                      }
                                                 )
          -- Then it's passed through the events
-         (sMap'', es') = resolveState sMap' es
-         
-         -- TODO onset is calculated in toOSC as well..
-         on e tempo'' = (sched tempo'' $ start $ wholeOrPart e)
-         (tes, tempo') = processCps tempo $ es'
-     forM_ cxs $ \cx@(Cx target _ oscs _ _) -> do
-         let latency = oLatency target + extraLatency
-             ms = concatMap (\(t, e) ->
-                              if (fake || (on e t) < frameEnd)
-                              then concatMap (toOSC latency busses e t) oscs
+        (sMap'', es') = resolveState sMap' es
+      tes <- processCps ops es'
+      -- For each OSC target
+      forM_ cxs $ \cx@(Cx target _ oscs _ _) -> do
+        -- Latency is configurable per target
+        -- We also add extra latency:
+        -- 0 for fake ticks
+        -- timeframe length + nudge for other ticks
+        let latency = oLatency target + extraLatency
+            ms = concatMap (\e ->
+                              -- All events in the playmap are added in case of a fake tick
+                              -- Q: Why do we need special handling for fake ticks? It seems to me like:
```

Comment:
No actual change requested here, but I am a bit worried there could be a regression here ref #501. It looks good though - I'll try it out!

Reply:
Interesting! However, it is not clear to me how the double counting happens. In tempo, we always use the end cycle as the start of next processing. cps does not change the cycle of the event, so the event should not be included in the query result, right? Does it have to do with event onsets?

I'll clean up this comment and add the issue number! Thanks 😃

Reply:
I think calculation shouldn't be done once per cycle, but at a fixed rate, currently by default 20 hertz.

Reply:
I believe I used confusing term "end cycle". What I meant refers to this code: https://github.com/tidalcycles/Tidal/blob/9753f73842944a3680d5584eab29c7e0c5fb06d1/src/Sound/Tidal/Tempo.hs#L197
The next arc's start cycle is the previous arc's end cycle. IF query only returns events that should be sent in this arc, then I don't see how this check is needed (since no cycle should be present in two different nowArcs unless there is a lot of drifting which is not what we are trying to solve here). IF query also returns events that shouldn't be sent in this arc (for example where the onset falls outside the arc but the tail does not), then I believe I understand why this is needed.

Reply:
The problem is that the frame length is in seconds. We convert from seconds to cycles to query the pattern, but then events can change the cps during the frame. That means we can end up with following events that are out of frame, hence this check. We can also end up missing events, we don't deal with that so far though..

Reply:
Ah, right! But why not send the events that are out of frame anyway? The events that are filtered out will never be added again in a later frame so they will unfortunately be dropped completely. Events that are missed in this frame will be picked up in the next frame, given that the next frame is scheduled quickly enough to not reach SuperCollider too late. The fixed rate is only used to set the end cycle of nowArc:
https://github.com/tidalcycles/Tidal/blob/9753f73842944a3680d5584eab29c7e0c5fb06d1/src/Sound/Tidal/Tempo.hs#L208-L214

Reply:
I see, hmm. So currently it seems that if the cps is slowed we might miss some events that get 'out of frame', but if it's speeded up all the events still make it within the latency buffer time.
Still, before this check was added, in-pattern cps changes could sound bad, I think with doubled up events. Feel free to remove it though, maybe the problem has been solved better elsewhere in the code.

Reply:
OK, then I'll remove it and if the issue reappears we know what to look for. Please ping me in such case and I'll dig deeper!